File: 0143-reorder-list/0143-reorder-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reorderList(self, head: Optional[ListNode]) -> None:
        def reverse(head):
            prev = None;
            while head != None:
                head.next, prev, head = prev, head, head.next
            return prev
        
        slow, fast = head, head
        while fast != None and fast.next != None:
            slow, fast = slow.next, fast.next.next
        
        if fast != None:   slow = slow.next
            
        fast = reverse(slow)
        slow = head
        
        while slow != None and fast != None:
            slow.next, fast.next, fast, slow = fast, slow.next, fast.next, slow.next
        
        if slow.next != None:   slow.next = None;

File: 0143-reorder-list/0143-reorder-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    private ListNode reverse(ListNode head) {
        ListNode prev = null;
        while (head != null) {
            ListNode next = head.next;
            head.next = prev;
            prev = head;
            head = next;
        }
        
        return prev;
    }
    
    public void reorderList(ListNode head) {
        ListNode slow = head, fast = head;
        while (fast != null && fast.next != null) {
            slow = slow.next;
            fast = fast.next.next;
        }
        
        if (fast != null)   slow = slow.next;
        fast = reverse(slow);
        
        slow = head;
        while (slow != null && fast != null) {
            ListNode next = slow.next;
            // add fast's node
            slow.next = fast;
            fast = fast.next;
            
            // link fast's node to slow's next
            slow = slow.next;
            slow.next = next;
            
            slow = next;
        }
        
        // to avoid self loop in odd-length case
        if (slow.next != null)   slow.next = null;
    }
}